### ISSUE TITLE
numerical error in Bayesian NN with HMC

### ISSUE THREAD

**Issue by zhaotianjing** [TuringLang/Turing.jl#1501]
Hi,

I followed the [Turing tutorial of Bayesian Neural Networks](https://turing.ml/dev/tutorials/3-bayesnn/) to implement a Bayesian Neural Network for prediction problem. In my BNN, the input layer contains a number of `L0` nodes, then there are `L1` nodes in the 1st hidden layer, `L2` nodes in the 2nd hidden layer, and one node in output layer. The number of individuals in my data is about `n=500`

I got the numerical error in every iteration when the number of unknows are huge (i.e., too many input nodes; too many hidden nodes). See below table:
| L0 | L1 | L2 |  |   
|----|----|----|-----------------|
|  15,000  | 2  |2   |  worked    | 
|  15,000  | 20   | 10   |  numerical error  | 
|   20,000 | 2   |  2  |  numerical error   | 
 

My BNN model is BNN with unknow variance components:
```
@model bayes_nn(x, y, L0, L1, L2) = begin
    v=4
    s=(v-2)/v
    # hyper-parameter
    varw0 ~ InverseGamma(v/2, v*s/2)  # = scale-inv-chisq(v,s)
    varw1 ~ InverseGamma(v/2, v*s/2)
    varw2 ~ InverseGamma(v/2, v*s/2)
    vare ~ InverseGamma(v/2, v*s/2)

    # weight&bias
    W0 ~ MvNormal(zeros(L0*L1), sqrt(varw0))
    b0 ~ MvNormal(zeros(L1), 1)
    W1 ~ MvNormal(zeros(L1*L2), sqrt(varw1))
    b1 ~ MvNormal(zeros(L2),1)
    W2 ~ MvNormal(zeros(L2), sqrt(varw2))
    b2 ~ Normal(0,1)

    # predictions
    preds = nn_forward(x, W0,b0,W1,b1,W2,b2,L0,L1,L2)   #1-by-n predicted value

    y ~ MvNormal(vec(preds), sqrt(vare))

end
```

The model is trained via:
```
ch = sample(bayes_nn(x_train, y_train,L0,L1,L2), HMC(0.1, 10), nIteration)
```

The error in every iteration is: 
```
┌ Warning: The current proposal will be rejected due to numerical error(s).
│   isfinite.((θ, r, ℓπ, ℓκ)) = (true, false, false, false)
└ @ AdvancedHMC ~/.julia/packages/AdvancedHMC/MIxdK/src/hamiltonian.jl:47
```

I found a[ similar posts for numerical error](https://discourse.julialang.org/t/numerical-errors-in-logit-normal-model-using-turing-jl/30602/2), and the numerical error in this post was caused by the unstable sigmoid function. 

So I guess the` InverseGamma` function used in my model is not numerically stable. So I fixed the hyper-parameter in my BNN as 1 to avoid using the `InverseGamma` function:
```
@model bayes_nn(x, y, L0, L1, L2) = begin

    # weight&bias
    W0 ~ MvNormal(zeros(L0*L1), 1)
    b0 ~ MvNormal(zeros(L1), 1)
    W1 ~ MvNormal(zeros(L1*L2), 1)
    b1 ~ MvNormal(zeros(L2),1)
    W2 ~ MvNormal(zeros(L2), 1)
    b2 ~ Normal(0,1)

    # predictions
    preds = nn_forward(x, W0,b0,W1,b1,W2,b2,L0,L1,L2)   #1-by-n predicted value

    y ~ MvNormal(vec(preds), 1)

end
```

I tested again and I don't have the numerical error! So it seems the ` InverseGamma` function is not numerically stable.

However, I still want to sample the variance component (i.e., hyperparameters) because I actually don't know the true values for them. Would you please tell me how to achieve this in Turing? 

Thank you,
Tianjing

**Comment by devmotion** [comment]
I haven't worked with BNN myself but my intuition would be to scale the variance with the number of hidden units since this is the "appropriate scaling" in the wide limit for recovering a Gaussian process (see, e.g., equation (6) in http://mlg.eng.cam.ac.uk/matthews/deepwide.pdf). I have mostly seen BNNs with fixed variance but maybe the scaling could resolve the instability that you observe.
